Resurrect BalForm (#1331)
balancer-labs/frontend-v2: src/components/_global/BalForm/readme.md

# BalForm

`<BalForm />` is a useful component for creating wrappers for simple and complex form structures on our frontend. There are a few motivations behind using a form component for creating forms:

> - Centralised and instance specific state. The form owns its state and can pass down its state (aka form values) to children at any nested level. This also means not needing to creating and managing your own form state in whichever component / composable that might need to feed into your form.

Reviewer: ```suggestion
- Centralised and instance-specific state. The form owns its state and can pass down its state (aka form values) to children at any nested level. This also means not needing to create and manage your own form state in whichever component / composable that might need to feed into your form.
```

- Centralised validation. If your component accepts in a rules prop, you can integrate it as a form component (more on this below) and have it validate against all passed in rules, whenever the components value changes. Any errors present during submission of the form will prevent the submission function from executing and a list of errors for each field be exported by the forms instance. 
- Simplified integration**. All you need to do is define the form component you are using as a child of `<BalForm />` to have its values, rules and any change handlers bound to the form, automatically.

  ** Assuming that all child form components (e.g. inputs, radio buttons, complex inputs such as <TokenInput/> have code that allows them to integrate as a form component.

## Usage

```typescript
// somecomponent.vue

/// script

// defines a component instance that will be
// passed into <BalForm />. Names are tied to 
// form instances and are unique. The value returned
// represents the form instance.
const form = useForm({ name: 'my-form' })

function mySubmit(formValues: T) {
 // do stuff
}

/// template

> <BalForm :form="form" :submit="form.handleSubmit(mySubmit)"

Reviewer: This is quite an unintuitive usage, I'd expect to just pass in a function to the submit prop, like:
```suggestion
<BalForm :form="form" :submit="mySubmit"
```
Is there no way we can get it down to this?

  <div v-if="form.errors.username.length > 0">There are errors in your username</div>
  <TokenInput name="tokenIn" />
  <input :value="form.register('username', [isLowerCase()])" />
</BalForm>
```

> In the above example. `<TokenInput />` is registered as a complex form component. However, all that needs to be done to register this component to the forms state is to simply give it a name and have it as a child of `<BalForm />`. 

Reviewer: I'd use the terminology 'wrapped input' here rather than 'complex component'
```suggestion
In the above example. `<TokenInput />` is registered as a wrapped form input. However, all that needs to be done to register this component to the forms state is to simply give it a name and have it as a child of `<BalForm />`. 
```

The other plain `<input />` component is not a complex component and displays how a plain html input component can be registered.

Any changes to values of either input component will be registered to the form instance and available to consume via `form.values`. Any validation errors that are registered from an inputs rules will be available via `form.errors`. 

Values which the form was submitted with will be passed to the `mySubmit` function, however this function will only be called IF there are no errors inside `form.errors`.

> ## Registering a complex component

Reviewer: ```suggestion
## Registering a wrapped input component
```

> In the above example the `<TokenInput />` component was registered as a complex component. The `TokenInput` component is comprised of two main parts. A token selector (address) and a normal test component (amount). 

Reviewer: ```suggestion
In the above example the `<TokenInput />` component was registered as a complex component. The `TokenInput` component is comprised of two main parts. A token selector (address) and a normal text component (amount). 
```
